File: android/app/src/main/kotlin/com/torvm/android/packet/UdpHeader.kt
package com.torvm.android.packet

/**
 * Parses and builds 8-byte UDP headers.
 *
 * All multi-byte fields use network byte order (big-endian).
 */
data class UdpHeader(
    val sourcePort: Int,
    val destPort: Int,
    val length: Int,
    val checksum: Int
) {
    /**
     * Serialize the UDP header to bytes.
     *
     * Note: The checksum field is written as-is. Callers should compute the
     * correct checksum using [Checksum.tcpUdpChecksum] after building the header.
     */
    fun toBytes(): ByteArray {
        val bytes = ByteArray(8)

        // Bytes 0-1: Source port
        bytes[0] = (sourcePort shr 8).toByte()
        bytes[1] = sourcePort.toByte()

        // Bytes 2-3: Destination port
        bytes[2] = (destPort shr 8).toByte()
        bytes[3] = destPort.toByte()

        // Bytes 4-5: Length
        bytes[4] = (length shr 8).toByte()
        bytes[5] = length.toByte()

        // Bytes 6-7: Checksum
        bytes[6] = (checksum shr 8).toByte()
        bytes[7] = checksum.toByte()

        return bytes
    }

    companion object {
        const val HEADER_SIZE = 8

        /**
         * Parse a UDP header from raw bytes.
         *
         * @param buffer the byte array containing the UDP datagram
         * @param offset the starting position of the UDP header within the buffer
         * @return a parsed UdpHeader
         * @throws IllegalArgumentException if the buffer is too short
         */
        fun parse(buffer: ByteArray, offset: Int): UdpHeader {
            require(buffer.size - offset >= 8) { "Buffer too short for UDP header: need 8 bytes, have ${buffer.size - offset}" }

            val sourcePort = ((buffer[offset].toInt() and 0xFF) shl 8) or
                (buffer[offset + 1].toInt() and 0xFF)

            val destPort = ((buffer[offset + 2].toInt() and 0xFF) shl 8) or
                (buffer[offset + 3].toInt() and 0xFF)

            val length = ((buffer[offset + 4].toInt() and 0xFF) shl 8) or
                (buffer[offset + 5].toInt() and 0xFF)

            val checksum = ((buffer[offset + 6].toInt() and 0xFF) shl 8) or
                (buffer[offset + 7].toInt() and 0xFF)

            return UdpHeader(
                sourcePort = sourcePort,
                destPort = destPort,
                length = length,
                checksum = checksum
            )
        }
    }
}

File: ios/TorVMTunnel/Packet/UdpHeader.swift
import Foundation

/// Parses and builds 8-byte UDP headers.
///
/// All multi-byte fields use network byte order (big-endian).
struct UdpHeader {
    static let headerSize = 8

    let sourcePort: UInt16
    let destPort: UInt16
    let length: UInt16
    let checksum: UInt16

    /// Serialize the UDP header to bytes.
    ///
    /// The checksum field is written as-is. Callers should compute the
    /// correct checksum using `Checksum.tcpUdpChecksum` after building the header.
    func toBytes() -> Data {
        var bytes = Data(count: 8)

        // Bytes 0-1: Source port
        bytes[0] = UInt8(sourcePort >> 8)
        bytes[1] = UInt8(sourcePort & 0xFF)

        // Bytes 2-3: Destination port
        bytes[2] = UInt8(destPort >> 8)
        bytes[3] = UInt8(destPort & 0xFF)

        // Bytes 4-5: Length
        bytes[4] = UInt8(length >> 8)
        bytes[5] = UInt8(length & 0xFF)

        // Bytes 6-7: Checksum
        bytes[6] = UInt8(checksum >> 8)
        bytes[7] = UInt8(checksum & 0xFF)

        return bytes
    }

    /// Parse a UDP header from raw bytes.
    ///
    /// - Parameters:
    ///   - data: the data containing the UDP datagram
    ///   - offset: the starting position of the UDP header within the data
    /// - Returns: a parsed UdpHeader
    /// - Throws: PacketError if the buffer is too short
    static func parse(data: Data, offset: Int) throws -> UdpHeader {
        guard data.count - offset >= 8 else {
            throw PacketError.bufferTooShort(needed: 8, have: data.count - offset)
        }

        let sourcePort = (UInt16(data[offset]) << 8) |
            UInt16(data[offset + 1])

        let destPort = (UInt16(data[offset + 2]) << 8) |
            UInt16(data[offset + 3])

        let length = (UInt16(data[offset + 4]) << 8) |
            UInt16(data[offset + 5])

        let checksum = (UInt16(data[offset + 6]) << 8) |
            UInt16(data[offset + 7])

        return UdpHeader(
            sourcePort: sourcePort,
            destPort: destPort,
            length: length,
            checksum: checksum
        )
    }
}
